hw/ip/kmac/data/kmac.hjson

```diff
@@ -0,0 +1,340 @@
+              name: "cSHAKE"
+              desc: "cSHAKE hashing mode. It appends `00` to the end of msg"
+            }
+          ]
+        } // f: mode
+        { bits: "8"
+          name: "msg_endianness"
+          desc: '''Message and Secret Key Endianness.
+
+                Convert TL-UL write data[31:0] to big-endian style
+                {d[7:0], d[15:8], ...}
+
+                0: Little-endian. Keep input value as it is
+                1: Big-endian. Convert incoming value
+                '''
+          resval: "1"
+        } // f: msg_endianness
+        { bits: "9"
+          name: "state_endianness"
+          desc: '''Output state (Digest) Endianness.
+
+                Convert read-out state register to big-endian style.
+                If 0, keep the internal value as it is when SW reads.
+                If 1, convert the internal state value.
+                '''
+        } // f: state_endianness
+      ]
+    } // R: CFG
+    { name: "CMD"
+      desc: '''KMAC/ SHA3 command register.
+
+            This register is to control the KMAC to start accepting message,
+            to process the message, and to manually run additional keccak
+            rounds at the end. Only at certain stage, the CMD affects to the
+            control logic. It follows the sequence of
+
+            `start` --> `process` --> {`run` if needed --> } `done`
+            '''
+      swaccess: "r0w1c"
+      hwaccess: "hro"
+      hwext: "true"
+      hwqe:  "true"
+      tags: [// design assertion : after start sets, can only wr msg or set process
+             // design assertion : process can be set only after start is set
+             "excl:CsrAllTests:CsrExclWrite"]
+      fields: [
+        { bits: "0"
+          name: "start"
+          desc: '''If writes 1 into this field when KMAC/SHA3 is in idle,
+                KMAC/SHA3 begins its operation.
+
+                If the mode is cSHAKE, before receiving the message, the
+                hashing logic processes Function name string N and
+                customization input string S first. If KMAC mode is enabled,
+                additionally it processes secret key block.
+                '''
+        } // f: start
+        { bits: "1"
+          name: "process"
+          desc: '''If writes 1 into this field when KMAC/SHA3 began its
+                operation and received the entire message, it computes the
+                digest or signing.
+                '''
+        } // f: process
+        { bits: "2"
+          name: "run"
+          desc: '''The `run` field is used in the sponge squeezing stage.
+                It triggers the keccak round logic to run full 24 rounds.
+                This is optional and used when software needs more digest bits
+                than the keccak rate.
+
+                It only affects when the kmac/sha3 operation is completed.
+                '''
+        } // f: run
+        { bits: "3"
+          name: "done"
```

Comment:
sorry @eunchan i think i'm confused about something.  I thought `done` was used to signal the completion of squeezing and not absorbing?

If that's the case, and done is used to clear the message fifo, shouldn't the message fifo be ready for a clear much earlier? sorry if i mixed things up, or if i'm remembering the wrong thing from `sha3core`.

Reply:
Yes that is right. Msg FIFO could be accept the next message right after process triggers internal SHA3 pad logic. But it is only around 72 cycles differences if no sponge squeezing happens. So I made the datapath simple.

Reply:
ah sounds good.  Should the description be updated to say write 1 to done when squeezing is completed? Is it supposed to say absorbing below?